doc/api/esm.md

```diff
@@ -945,20 +948,40 @@ validating the import assertion.
 
 The final value of `format` must be one of the following:
 
-| `format`     | Description                    | Acceptable types for `source` returned by `load`      |
-| ------------ | ------------------------------ | ----------------------------------------------------- |
-| `'builtin'`  | Load a Node.js builtin module  | Not applicable                                        |
-| `'commonjs'` | Load a Node.js CommonJS module | Not applicable                                        |
-| `'json'`     | Load a JSON file               | { [`string`][], [`ArrayBuffer`][], [`TypedArray`][] } |
-| `'module'`   | Load an ES module              | { [`string`][], [`ArrayBuffer`][], [`TypedArray`][] } |
-| `'wasm'`     | Load a WebAssembly module      | { [`ArrayBuffer`][], [`TypedArray`][] }               |
+| `format`     | Description                    | Acceptable types for `source` returned by `load`                           |
+| ------------ | ------------------------------ | -------------------------------------------------------------------------- |
+| `'builtin'`  | Load a Node.js builtin module  | Not applicable                                                             |
+| `'commonjs'` | Load a Node.js CommonJS module | { [`string`][], [`ArrayBuffer`][], [`TypedArray`][], `null`, `undefined` } |
+| `'json'`     | Load a JSON file               | { [`string`][], [`ArrayBuffer`][], [`TypedArray`][] }                      |
+| `'module'`   | Load an ES module              | { [`string`][], [`ArrayBuffer`][], [`TypedArray`][] }                      |
+| `'wasm'`     | Load a WebAssembly module      | { [`ArrayBuffer`][], [`TypedArray`][] }                                    |
 
 The value of `source` is ignored for type `'builtin'` because currently it is
-not possible to replace the value of a Node.js builtin (core) module. The value
-of `source` is ignored for type `'commonjs'` because the CommonJS module loader
-does not provide a mechanism for the ES module loader to override the
-[CommonJS module return value](#commonjs-namespaces). This limitation might be
-overcome in the future.
+not possible to replace the value of a Node.js builtin (core) module.
+
+For type `'commonjs'`, if `source` is defined, `require` calls from inside the
+loaded module will go through the registered `resolve` and `load` hooks.
+If `source` is undefined or `null`, it will be handled by the CommonJS
+module loader and not processed by any registered hooks. This behavior
+for nullish `source` is temporary; in the future, a `source` value that is
+undefined or `null` will not be supported.
```

Comment:
nit (the first bit reads a little yoda-esque), and I think it's worth specifically calling out CJSLoader won't be used and require.extensions won't apply.

```suggestion
If a `load` hook returns commonjs `source`, that module's `require` calls will
be processed by ESM loader with registered `resolve` and `load` hooks;
`require.extensions` and monkey-patching on the CommonJS module loader will not
apply. If `source` is undefined or `null`, it will be handled by the CommonJS
module loader and not processed by registered hooks. This behavior for nullish
`source` is temporary—in the future, nullish `source` will not be supported.
```